module/support/Gazebo/src/Gazebo.cpp

```diff
@@ -0,0 +1,360 @@
+
+            connected = false;
+
+            // Set up transport node for joint control
+            // This will be ADVERTISED to the Ctrl topic
+            ignition::transport::NodeOptions joint_ctrl_node_opts;
+            joint_ctrl_node_opts.SetPartition(config["partition"]["joint"]);
+            joint_ctrl_node_opts.SetNameSpace(config["ignition_namespace"]);
+            joint_ctrl = std::make_unique<ignition::transport::Node>(joint_ctrl_node_opts);
+
+            // Set up transport node for joint status
+            // This will be SUBSCRIBED to the Status topic
+            ignition::transport::NodeOptions joint_status_node_opts;
+            joint_status_node_opts.SetPartition(config["partition"]["joint"]);
+            joint_status_node_opts.SetNameSpace(config["ignition_namespace"]);
+            joint_status = std::make_unique<ignition::transport::Node>(joint_status_node_opts);
+
+            // Set up transport node for world control
+            // This will be ADVERTISED to the Ctrl topic
+            ignition::transport::NodeOptions world_ctrl_node_opts;
+            world_ctrl_node_opts.SetPartition(config["partition"]["world"]);
+            world_ctrl_node_opts.SetNameSpace(config["ignition_namespace"]);
+            world_ctrl = std::make_unique<ignition::transport::Node>(world_ctrl_node_opts);
+
+            // Set up transport node for world status
+            // This will be SUBSCRIBED to the Status topic
+            ignition::transport::NodeOptions world_status_node_opts;
+            world_status_node_opts.SetPartition(config["partition"]["world"]);
+            world_status_node_opts.SetNameSpace(config["ignition_namespace"]);
+            world_status = std::make_unique<ignition::transport::Node>(world_status_node_opts);
+
+            // Set up transport node for ball status
+            // This will be SUBSCRIBED to the Status topic
+            ignition::transport::NodeOptions ball_status_node_opts;
+            ball_status_node_opts.SetPartition(config["partition"]["ball"]);
+            ball_status_node_opts.SetNameSpace(config["ignition_namespace"]);
+            ball_status = std::make_unique<ignition::transport::Node>(ball_status_node_opts);
+
+            auto ad_msg_opts = std::make_unique<ignition::transport::AdvertiseMessageOptions>();
+
+            discovery_node = std::make_unique<ignition::transport::MsgDiscovery>(p_uuid, gazebo_message_port);
+
+            msg_publisher = std::make_unique<ignition::transport::MessagePublisher>(
+                topic_ctrl, host_addr, ctrl_addr, p_uuid, n_uuid, "ADVERTISE", *ad_msg_opts);
+
+            msg_pub_world = std::make_unique<ignition::transport::MessagePublisher>(
+                topic_world_ctrl, host_addr, ctrl_addr, p_uuid, n_uuid, "ADVERTISE", *ad_msg_opts);
+
+            bool joint_status_cb =
+                joint_status->Subscribe<Gazebo, ignition::msgs::StringMsg>(topic_status, &Gazebo::JointStatusCB, this);
```

Comment:
Second argument here is a protocol buffer type. So to swap over to using custom protocol buffers instead of using strings and parsing them, you can use this pattern.

```c++
joint_status->Subscribe<Gazebo, ::protobuf::message::input::SomeMessage>(topic_status, &Gazebo::JointStatus, this);
...
void Gazebo::JointStatus(const message::input::SomeMessage& msg) {
...
}
```
This should work as neutron stores all the messages in the same namespace with ::protobuf on the start, and those protocol buffers are castable to the Neutron types.